File: HackerRank/Day0/Day0-MeanMedianMode.py
# Jeremiah Flaga
# July 13, 2017

dict = {}

num_of_items = int(input())
list_of_numbers = list(map(int, input().split()))

for item in list_of_numbers:
    if item in dict:
        dict[item] += 1
    else:
        dict[item] = 1

#for k in dict.keys():
#    print('{}: {}'.format(k, dict[k]))

sum = 0
item_with_max_frequency = 0
max_frequency = 0
for item,freq in dict.items():
    sum += item * freq
    if freq == max_frequency and item < item_with_max_frequency:
        item_with_max_frequency = item
    elif freq > max_frequency:
        max_frequency = freq
        item_with_max_frequency = item
    
mean = sum / num_of_items
mode = item_with_max_frequency

# computing for median
list_of_numbers.sort()
median = 0
if num_of_items % 2 == 0:
    middle_item_1 = list_of_numbers[(num_of_items // 2) - 1]
    middle_item_2 = list_of_numbers[num_of_items // 2]
    median = (middle_item_1 + middle_item_2) / 2
else:
    median = middle_item_2 = list_of_numbers[num_of_items // 2]

print(mean)
print(median)
print(mode)

File: HackerRank/10DaysOfStatistics/Day0-MeanMedianMode/Day0-MeanMedianMode.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <iomanip>
using namespace std;

/***
# Jeremiah Flaga
# July 10, 2017
*/

int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
    long nItems = 0;
    cin >> nItems;
    
    if (nItems > 0) {
        vector<long> items(nItems);
        map<long, long> mapOfItemToNumberOfOccurrences;
        double sum = 0;
        long maxNumOfOccurences = 0;
        long mode = 0;

        for (long i = 0; i < nItems; i++) {
            long item;;
            cin >> item;
            items[i] = item;

            sum += item;

            mapOfItemToNumberOfOccurrences[item]++;
            if (mapOfItemToNumberOfOccurrences[item] > maxNumOfOccurences) {
                maxNumOfOccurences = mapOfItemToNumberOfOccurrences[item];
                mode = item;
            } else if (mapOfItemToNumberOfOccurrences[item] == maxNumOfOccurences && item < mode) {
                    mode = item;
            }
        }

        cout << setprecision(1) << fixed;

        //mean
        double mean = sum/nItems;
        cout << sum/nItems << endl;

        //median
        double median = 0.0;
        sort(items.begin(), items.end());
        bool isNumberOfItemsOdd = nItems % 2 != 0;
        if (isNumberOfItemsOdd)
            median = items[(nItems/2)];
        else
            median = (items[(nItems/2) - 1]  + items[(nItems/2)]) / 2.0;

        cout << median << endl;

        //mode
        cout << mode;
    }
}
